build.gradle

```diff
@@ -31,23 +28,36 @@ processResources {
 	}
 }
 
-// ensure that the encoding is set to UTF-8, no matter what the system default is
-// this fixes some edge cases with special characters not displaying correctly
-// see http://yodaconditions.net/blog/fix-for-java-file-encoding-problems-with-gradle.html
-tasks.withType(JavaCompile) {
-	options.encoding = "UTF-8"
+tasks.withType(JavaCompile).configureEach {
+	// ensure that the encoding is set to UTF-8, no matter what the system default is
+	// this fixes some edge cases with special characters not displaying correctly
+	// see http://yodaconditions.net/blog/fix-for-java-file-encoding-problems-with-gradle.html
+	// If Javadoc is generated, this must be specified in that task too.
+	it.options.encoding = "UTF-8"
+
+	// The Minecraft launcher currently installs Java 8 for users, so your mod probably wants to target Java 8 too
+	// JDK 9 introduced a new way of specifying this that will make sure no newer classes or methods are used.
+	// We'll use that if it's available, but otherwise we'll use the older option.
+	def targetVersion = JavaVersion.VERSION_1_8
+	if (JavaVersion.current().isJava9Compatible()) {
+		 it.options.release = targetVersion.ordinal() + 1
```

Comment:
This + 1 is a little odd looking, just hard coding it to 8 might be clearer

Reply:
Yeah, it's a bit of an unfortunate workaround to the fact that [`JavaVersion.getMajorVersion()`](https://github.com/gradle/gradle/blob/46a0779dabbd2d7518cc6ae1bcae1030d4ea1100/subprojects/base-services/src/main/java/org/gradle/api/JavaVersion.java#L269) returns a String of the same value, and not an int. If only Gradle didn't have 3 separate ways to represent a Java version in different parts of its API....

Can probably avoid it though by declaring `targetVersion` as an int, and converting that to `JavaVersion` for the pre-Java9 variant.